### PR TITLE
Fix intermediate header in aggregate projection

### LINKED ISSUE #37151: Logical error: Cannot find column in projection but query analysis still selects this projection
https://s3.amazonaws.com/clickhouse-test-reports/0/2a4f763a5ea752bb187c8304a1b42c5b5a951a47/fuzzer_astfuzzerubsan,actions//report.html

```
2022.05.12 14:20:26.276674 [ 173 ] {d6d49119-26e3-43e3-8ff1-07c4a56cb3ec} <Fatal> : Logical error: 'Cannot find column tupleElement(_partition_value, 1) in minmax_count projection but query analysis still selects this projection. It's a bug'.
2022.05.12 14:21:58.581126 [ 679 ] {} <Fatal> BaseDaemon: ########################################
2022.05.12 14:21:58.581167 [ 679 ] {} <Fatal> BaseDaemon: (version 22.5.1.1 (official build), build id: 44A8856E094F5335) (from thread 173) (query_id: d6d49119-26e3-43e3-8ff1-07c4a56cb3ec) (query: SELECT min(i), max(i), count() FROM d WHERE (_partition_value.1) = 0 GROUP BY ignore(bitTest(ignore(NULL), 65535), NULL, (_partition_value.1) = 7, '10.25', bitTest(NULL, -9223372036854775808), NULL, ignore(ignore(-2147483647, NULL)), 1024), _partition_id ORDER BY _partition_id ASC NULLS FIRST) Received signal Aborted (6)
2022.05.12 14:21:58.581193 [ 679 ] {} <Fatal> BaseDaemon: 
2022.05.12 14:21:58.581240 [ 679 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f31d327903b 0x7f31d3258859 0xf7ab9e0 0xf7abce2 0xf89fb19 0x2377a7a0 0x2377dee2 0x2377c030 0x2377f211 0x2283b66e 0x2282f2b5 0x228286c0 0x22824dbc 0x228a19dc 0x2289ea87 0x2289c801 0x227da9e8 0x227d943e 0x22c9fdab 0x22c9d18a 0x23c44e78 0x23c61e96 0x25aef46c 0x25aef945 0x25c6d507 0x25c6adec 0x7f31d3430609 0x7f31d3355163
2022.05.12 14:21:58.581283 [ 679 ] {} <Fatal> BaseDaemon: 3. gsignal @ 0x7f31d327903b in ?
2022.05.12 14:21:58.581304 [ 679 ] {} <Fatal> BaseDaemon: 4. abort @ 0x7f31d3258859 in ?
2022.05.12 14:21:58.614112 [ 679 ] {} <Fatal> BaseDaemon: 5. ./build_docker/../src/Common/Exception.cpp:48: DB::handle_error_code(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool, std::__1::vector<void*, std::__1::allocator<void*> > const&) @ 0xf7ab9e0 in /workspace/clickhouse
2022.05.12 14:21:58.644866 [ 679 ] {} <Fatal> BaseDaemon: 6. ./build_docker/../src/Common/Exception.cpp:0: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xf7abce2 in /workspace/clickhouse
2022.05.12 14:21:58.668994 [ 679 ] {} <Fatal> BaseDaemon: 7.1. inlined from ./build_docker/../contrib/libcxx/include/string:1445: std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__is_long() const
2022.05.12 14:21:58.669043 [ 679 ] {} <Fatal> BaseDaemon: 7.2. inlined from ../contrib/libcxx/include/string:2231: ~basic_string
2022.05.12 14:21:58.669067 [ 679 ] {} <Fatal> BaseDaemon: 7. ../src/Common/Exception.h:39: DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(int, fmt::v8::basic_format_string<char, fmt::v8::type_identity<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>::type>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0xf89fb19 in /workspace/clickhouse
2022.05.12 14:21:58.878260 [ 679 ] {} <Fatal> BaseDaemon: 8. ./build_docker/../src/Storages/MergeTree/MergeTreeData.cpp:5284: DB::MergeTreeData::getMinMaxCountProjectionBlock(std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, bool, DB::SelectQueryInfo const&, std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const> > > const&, std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const> > >&, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, long, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, long> > > const*, std::__1::shared_ptr<DB::Context const>) const @ 0x2377a7a0 in /workspace/clickhouse
2022.05.12 14:21:59.047800 [ 679 ] {} <Fatal> BaseDaemon: 9. ./build_docker/../src/Storages/MergeTree/MergeTreeData.cpp:0: DB::MergeTreeData::getQueryProcessingStageWithAggregateProjection(std::__1::shared_ptr<DB::Context const>, std::__1::shared_ptr<DB::StorageSnapshot const> const&, DB::SelectQueryInfo&) const::$_27::operator()() const @ 0x2377dee2 in /workspace/clickhouse
2022.05.12 14:21:59.215285 [ 679 ] {} <Fatal> BaseDaemon: 10. ./build_docker/../src/Storages/MergeTree/MergeTreeData.cpp:5661: DB::MergeTreeData::getQueryProcessingStageWithAggregateProjection(std::__1::shared_ptr<DB::Context const>, std::__1::shared_ptr<DB::StorageSnapshot const> const&, DB::SelectQueryInfo&) const @ 0x2377c030 in /workspace/clickhouse
2022.05.12 14:21:59.386019 [ 679 ] {} <Fatal> BaseDaemon: 11.1. inlined from ./build_docker/../contrib/libcxx/include/optional:321: std::__1::__optional_storage_base<DB::ProjectionCandidate, false>::has_value() const
2022.05.12 14:21:59.386064 [ 679 ] {} <Fatal> BaseDaemon: 11.2. inlined from ../contrib/libcxx/include/optional:975: std::__1::optional<DB::ProjectionCandidate>::operator bool() const
2022.05.12 14:21:59.386081 [ 679 ] {} <Fatal> BaseDaemon: 11. ../src/Storages/MergeTree/MergeTreeData.cpp:5760: DB::MergeTreeData::getQueryProcessingStage(std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, std::__1::shared_ptr<DB::StorageSnapshot const> const&, DB::SelectQueryInfo&) const @ 0x2377f211 in /workspace/clickhouse
2022.05.12 14:21:59.473944 [ 679 ] {} <Fatal> BaseDaemon: 12. ./build_docker/../src/Interpreters/InterpreterSelectQuery.cpp:668: DB::InterpreterSelectQuery::getSampleBlockImpl() @ 0x2283b66e in /workspace/clickhouse
2022.05.12 14:21:59.556888 [ 679 ] {} <Fatal> BaseDaemon: 13. ./build_docker/../src/Interpreters/InterpreterSelectQuery.cpp:552: DB::InterpreterSelectQuery::InterpreterSelectQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context const>, std::__1::optional<DB::Pipe>, std::__1::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::SubqueryForSet, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, DB::SubqueryForSet> > >, std::__1::unordered_map<DB::PreparedSetKey, std::__1::shared_ptr<DB::Set>, DB::PreparedSetKey::Hash, std::__1::equal_to<DB::PreparedSetKey>, std::__1::allocator<std::__1::pair<DB::PreparedSetKey const, std::__1::shared_ptr<DB::Set> > > >)::$_1::operator()(bool) const @ 0x2282f2b5 in /workspace/clickhouse
2022.05.12 14:21:59.636784 [ 679 ] {} <Fatal> BaseDaemon: 14. ./build_docker/../src/Interpreters/InterpreterSelectQuery.cpp:0: DB::InterpreterSelectQuery::InterpreterSelectQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context const>, std::__1::optional<DB::Pipe>, std::__1::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::SubqueryForSet, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, DB::SubqueryForSet> > >, std::__1::unordered_map<DB::PreparedSetKey, std::__1::shared_ptr<DB::Set>, DB::PreparedSetKey::Hash, std::__1::equal_to<DB::PreparedSetKey>, std::__1::allocator<std::__1::pair<DB::PreparedSetKey const, std::__1::shared_ptr<DB::Set> > > >) @ 0x228286c0 in /workspace/clickhouse
2022.05.12 14:21:59.716634 [ 679 ] {} <Fatal> BaseDaemon: 15. ./build_docker/../src/Interpreters/InterpreterSelectQuery.cpp:165: DB::InterpreterSelectQuery::InterpreterSelectQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context const>, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0x22824dbc in /workspace/clickhouse
2022.05.12 14:21:59.749606 [ 679 ] {} <Fatal> BaseDaemon: 16. ./build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:725: std::__1::__unique_if<DB::InterpreterSelectQuery>::__unique_single std::__1::make_unique<DB::InterpreterSelectQuery, std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context>&, DB::SelectQueryOptions&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&>(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context>&, DB::SelectQueryOptions&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0x228a19dc in /workspace/clickhouse
2022.05.12 14:21:59.776115 [ 679 ] {} <Fatal> BaseDaemon: 17.1. inlined from ./build_docker/../contrib/libcxx/include/__memory/compressed_pair.h:48: __compressed_pair_elem<DB::InterpreterSelectQuery *, void>
2022.05.12 14:21:59.776143 [ 679 ] {} <Fatal> BaseDaemon: 17.2. inlined from ../contrib/libcxx/include/__memory/compressed_pair.h:130: __compressed_pair<DB::InterpreterSelectQuery *, std::__1::default_delete<DB::InterpreterSelectQuery> >
2022.05.12 14:21:59.776160 [ 679 ] {} <Fatal> BaseDaemon: 17.3. inlined from ../contrib/libcxx/include/__memory/unique_ptr.h:220: unique_ptr<DB::InterpreterSelectQuery, std::__1::default_delete<DB::InterpreterSelectQuery>, void, void>
2022.05.12 14:21:59.776176 [ 679 ] {} <Fatal> BaseDaemon: 17. ../src/Interpreters/InterpreterSelectWithUnionQuery.cpp:223: DB::InterpreterSelectWithUnionQuery::buildCurrentChildInterpreter(std::__1::shared_ptr<DB::IAST> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0x2289ea87 in /workspace/clickhouse
2022.05.12 14:21:59.801867 [ 679 ] {} <Fatal> BaseDaemon: 18. ./build_docker/../src/Interpreters/InterpreterSelectWithUnionQuery.cpp:0: DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context const>, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0x2289c801 in /workspace/clickhouse
2022.05.12 14:21:59.813537 [ 679 ] {} <Fatal> BaseDaemon: 19. ./build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:0: std::__1::__unique_if<DB::InterpreterSelectWithUnionQuery>::__unique_single std::__1::make_unique<DB::InterpreterSelectWithUnionQuery, std::__1::shared_ptr<DB::IAST>&, std::__1::shared_ptr<DB::Context>&, DB::SelectQueryOptions const&>(std::__1::shared_ptr<DB::IAST>&, std::__1::shared_ptr<DB::Context>&, DB::SelectQueryOptions const&) @ 0x227da9e8 in /workspace/clickhouse
2022.05.12 14:21:59.823763 [ 679 ] {} <Fatal> BaseDaemon: 20.1. inlined from ./build_docker/../contrib/libcxx/include/__memory/compressed_pair.h:48: __compressed_pair_elem<DB::InterpreterSelectWithUnionQuery *, void>
2022.05.12 14:21:59.823790 [ 679 ] {} <Fatal> BaseDaemon: 20.2. inlined from ../contrib/libcxx/include/__memory/compressed_pair.h:130: __compressed_pair<DB::InterpreterSelectWithUnionQuery *, std::__1::default_delete<DB::InterpreterSelectWithUnionQuery> >
2022.05.12 14:21:59.823805 [ 679 ] {} <Fatal> BaseDaemon: 20.3. inlined from ../contrib/libcxx/include/__memory/unique_ptr.h:220: unique_ptr<DB::InterpreterSelectWithUnionQuery, std::__1::default_delete<DB::InterpreterSelectWithUnionQuery>, void, void>
2022.05.12 14:21:59.823820 [ 679 ] {} <Fatal> BaseDaemon: 20. ../src/Interpreters/InterpreterFactory.cpp:122: DB::InterpreterFactory::get(std::__1::shared_ptr<DB::IAST>&, std::__1::shared_ptr<DB::Context>, DB::SelectQueryOptions const&) @ 0x227d943e in /workspace/clickhouse
2022.05.12 14:21:59.865510 [ 679 ] {} <Fatal> BaseDaemon: 21. ./build_docker/../src/Interpreters/executeQuery.cpp:658: DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x22c9fdab in /workspace/clickhouse
2022.05.12 14:21:59.910680 [ 679 ] {} <Fatal> BaseDaemon: 22. ./build_docker/../src/Interpreters/executeQuery.cpp:1067: DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x22c9d18a in /workspace/clickhouse
2022.05.12 14:21:59.947672 [ 679 ] {} <Fatal> BaseDaemon: 23. ./build_docker/../src/Server/TCPHandler.cpp:332: DB::TCPHandler::runImpl() @ 0x23c44e78 in /workspace/clickhouse
2022.05.12 14:21:59.993847 [ 679 ] {} <Fatal> BaseDaemon: 24. ./build_docker/../src/Server/TCPHandler.cpp:1783: DB::TCPHandler::run() @ 0x23c61e96 in /workspace/clickhouse
2022.05.12 14:21:59.997768 [ 679 ] {} <Fatal> BaseDaemon: 25. ./build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x25aef46c in /workspace/clickhouse
2022.05.12 14:22:00.004122 [ 679 ] {} <Fatal> BaseDaemon: 26.1. inlined from ./build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:54: std::__1::default_delete<Poco::Net::TCPServerConnection>::operator()(Poco::Net::TCPServerConnection*) const
2022.05.12 14:22:00.004151 [ 679 ] {} <Fatal> BaseDaemon: 26.2. inlined from ../contrib/libcxx/include/__memory/unique_ptr.h:315: std::__1::unique_ptr<Poco::Net::TCPServerConnection, std::__1::default_delete<Poco::Net::TCPServerConnection> >::reset(Poco::Net::TCPServerConnection*)
2022.05.12 14:22:00.004167 [ 679 ] {} <Fatal> BaseDaemon: 26.3. inlined from ../contrib/libcxx/include/__memory/unique_ptr.h:269: ~unique_ptr
2022.05.12 14:22:00.004182 [ 679 ] {} <Fatal> BaseDaemon: 26. ../contrib/poco/Net/src/TCPServerDispatcher.cpp:116: Poco::Net::TCPServerDispatcher::run() @ 0x25aef945 in /workspace/clickhouse
2022.05.12 14:22:00.010752 [ 679 ] {} <Fatal> BaseDaemon: 27. ./build_docker/../contrib/poco/Foundation/src/ThreadPool.cpp:213: Poco::PooledThread::run() @ 0x25c6d507 in /workspace/clickhouse
2022.05.12 14:22:00.017207 [ 679 ] {} <Fatal> BaseDaemon: 28.1. inlined from ./build_docker/../contrib/poco/Foundation/include/Poco/SharedPtr.h:156: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable> >::assign(Poco::Runnable*)
2022.05.12 14:22:00.017236 [ 679 ] {} <Fatal> BaseDaemon: 28.2. inlined from ../contrib/poco/Foundation/include/Poco/SharedPtr.h:208: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable> >::operator=(Poco::Runnable*)
2022.05.12 14:22:00.017252 [ 679 ] {} <Fatal> BaseDaemon: 28. ../contrib/poco/Foundation/src/Thread_POSIX.cpp:360: Poco::ThreadImpl::runnableEntry(void*) @ 0x25c6adec in /workspace/clickhouse
2022.05.12 14:22:00.017271 [ 679 ] {} <Fatal> BaseDaemon: 29. ? @ 0x7f31d3430609 in ?
2022.05.12 14:22:00.017288 [ 679 ] {} <Fatal> BaseDaemon: 30. __clone @ 0x7f31d3355163 in ?
2022.05.12 14:22:00.295734 [ 679 ] {} <Fatal> BaseDaemon: Checksum of the binary: 2CC6D314D57C9394B3F6564A74B0B337, integrity check passed.


```

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix projection exception when aggregation keys are wrapped inside other functions. This fixes https://github.com/ClickHouse/ClickHouse/issues/37151

The bug was introduced in https://github.com/ClickHouse/ClickHouse/pull/35788 , which tries to fix another issue incorrectly. This PR reverts the fix and replace it with the correct one.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -5385,7 +5385,7 @@ std::optional<ProjectionCandidate> MergeTreeData::getQueryProcessingStageWithAgg
         if (analysis_result.before_where)
         {
             candidate.where_column_name = analysis_result.where_column_name;
-            candidate.remove_where_filter = analysis_result.remove_where_filter;
+            candidate.remove_where_filter = !required_columns.contains(analysis_result.where_column_name);
```
Comment (amosbird) [inline src/Storages/MergeTree/MergeTreeData.cpp:5388]: Here is the fix which should be used in https://github.com/ClickHouse/ClickHouse/pull/35788

**Review by novikd — APPROVED** [review]

**Comment by novikd** [comment]
@Mergifyio update

**Comment by novikd** [comment]
@Mergifyio update

**Comment by novikd** [comment]
@Mergifyio update

**Comment by novikd** [comment]
@amosbird It seems like AST fuzzer has found a bug in this fix.

**Comment by amosbird** [comment]
@novikd I think it's alright now.

**Review by novikd — APPROVED** [review]

**Comment by novikd** [comment]
> Stateless tests (thread, actions) [3/3] — fail: 1, passed: 1384, skipped: 9

Looks unrelated.

**Pull request merged**